### PR TITLE
[EDA-804] Adding logic to copy logs even on error

### PR DESCRIPTION
**Pull request by skyler-rs**
> ### Motivate of the pull request
> - [x] To address an existing issue. If so, please provide a link to the issue: EDA-804
> - [ ] Breaking new feature. If so, please describe details in the description part.

> ### Describe the technical details
> #### What is currently done? (Provide issue link if applicable)
> - Logs were copied/renamed at the end of a given Task.
>    - If the task returned early, the logs would not get copied
>
> #### What does this pull request change?
> - Adding open source Scope Guard library to allow for firing code when exiting scope
>    - https://github.com/ricab/scope_guard
>  - Putting all CompilerOpengFPGA.cpp log copies into scope guards to ensure the logs get copied if they exist.

> ### Which part of the code base require a change
> 
> - [x] Library: Compiler
> - [ ] Plug-in: <Specify the plugin name>
> - [ ] Engine
> - [ ] Documentation
> - [ ] Regression tests
> - [ ] Continous Integration (CI) scripts

> ### Testing
> - Manually tested in Raptor using testcase listed in EDA-804 description

### CODE REVIEW

**Review by KochynVolodymyr — APPROVED** [review]

File: src/Compiler/CompilerOpenFPGA.cpp [hunk 1]
```diff
@@ -44,6 +44,7 @@
 #include "NewProject/ProjectManager/project_manager.h"
 #include "Utils/FileUtils.h"
 #include "Utils/LogUtils.h"
+#include "Utils/ScopeGuard.hpp"
```
Comment (ravic-rs) [inline src/Compiler/CompilerOpenFPGA.cpp:47]: Why is it not in third_party ?
https://github.com/os-fpga/FOEDAG/tree/main/third_party

File: src/Compiler/CompilerOpenFPGA.cpp [hunk 2]
```diff
@@ -1034,6 +1035,15 @@ bool CompilerOpenFPGA::Analyze() {
 }
 
 bool CompilerOpenFPGA::Synthesize() {
+  SCOPE_EXIT {
```
Comment (ravic-rs) [inline src/Compiler/CompilerOpenFPGA.cpp:1038]: https://en.cppreference.com/w/cpp/experimental/scope_exit

File: src/Compiler/CompilerOpenFPGA.cpp (hunk 1 above)
Comment (skyler-rs) [reply src/Compiler/CompilerOpenFPGA.cpp:47]: I was just thinking about it as a util and third party didn't occur to me. I'll move it.

File: src/Compiler/CompilerOpenFPGA.cpp (hunk 2 above)
Comment (skyler-rs) [reply src/Compiler/CompilerOpenFPGA.cpp:1038]: I saw that, but experimental didn't sound like a good title to be using in production code.

**Comment by skyler-rs** [comment]
Noticed that Analysis task doesn't get headers added to it's analysis.rpt log in CompilerOpenFPGA.cpp because it uses a slightly different method for logging. I've updated the code to add headers to this file as well.

**Review by KochynVolodymyr — APPROVED** [review]
Looks good!

**Comment by ravic-rs** [comment]
Description needs an update for change in scope guard source.

**Comment by skyler-rs** [comment]
> Description needs an update for change in scope guard source.

Fixed

**Pull request merged**